---
title: "Workshop Mutation PSPG 245B"
author: "Lee and Martell"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE )
```

# PSPG 245B
In this workshop you will learn about the fundamentals of how to download and utilize cancer genomic data. By the end of the course you should be able to generate your own full R markdown for your assigned TCGA ID, including plots and identification of actionable mutations. 

  * Please see the the accompanying PDF for more information.


## The following are required for this workshop
<div style="background-color:#b6d2db">

Software requirements  

* R >3.5 and have the below packages installed. 
* use BiocManager::install() when possible. 
    + cgdsr
    + ggplot2
    + knitr
    + kableExtra
    + dplyr
    + VennDiagram
    + reshape2
    + gridExtra
    + ggrepel
    
Please make sure to either clone the github repo or download the entire dropbox folder. Unzip this to a folder.  

* https://www.dropbox.com/sh/lvaqmurdozncnxw/AADWRtK8woO4MjXIvGqcUcpXa?dl=1

* https://github.com/ahdee/workshop.PSPG-245B.2.git

In this workshop we will be mainly using cbioportal's API to download the data we need. Thus it is important that you have access to the internet.

We will be using the TCGA study on Breast Invasive Carcinoma (TCGA, Cell 2015) 

* Each student should had been assigned a TCGA ID prior to this.
    + TCGA-C8-A3M7-01
    + TCGA-GM-A2D9-01
    + TCGA-BH-A1FL-01
    + TCGA-E2-A14Z-01
	+ TCGA-BH-A1FC-01
	+ TCGA-AC-A2QJ-01
	+ TCGA-EW-A1P8-01
	+ TCGA-E2-A1LE-01
	+ TCGA-E2-A1LK-01
	+ TCGA-AC-A2FE-01
	
</div>

<hr>

<font color="red"> Important: please remember to set the working directory to your source by clicking Sesssion-> Set working directory -> to source file location</font>

```{r, include=T, echo=T, message=FALSE, warning=FALSE, fig=TRUE}

# load require package and "source" an auxillary R script design for this workshop.  
# the auxillary script will help manage some of the more complex/redundant tasks 

library ( cgdsr)
library ( ggplot2)
library(knitr)
library(kableExtra)
library ( dplyr)
library ( reshape2)
library(VennDiagram)
library(gridExtra)
library (ggrepel)
# make sure to set to working directory 
# Session->set working directory -> to source file location 
source("auxi.R")

# initiate cbioportal API

cbiop = CGDS("http://www.cbioportal.org/")
```

# Lets learn how communicate with cbioportal and study what is available. 
* This portal is hierarchically structured. This is the workflow will will use. 
    + List of all the cancer studies available on this portal. 
    + Identify the study you want
    + List all the "cases" for that study, a case is a description of both data available for the study and/or subsets of certain phenotypes
	+ With each case you can get the list of all the associated TCGA identify
	+ which you can then use to get all the clincal data assocaited with the TCGA samples. 
	+ moreover you use the study to get specific genetic profile ( datasets available for downloading ). 
	+ From here you can then download the actual data. 

```{r }

# Start by getting a list of cancer studies available at cbio
studies = getCancerStudies(cbiop)

# This will return a table with a list of available cancer studies to draw from 
k2( head ( studies, 10) , "example list of data available in cbioportal")

# note: k2 is one of those cheat codes to make tables print out nicely 
# ? what is the "10" for in the function above. 
# printing the top ten rows

# lets find out how many studies are available to the public
dim ( studies ) # dim is a function to tabulate a dataframe

# ? how do you think we can use this to look for the dataset we need among > 200 studies?
# index or filter for the row which has a matching name to the desired dataset

# Find the dataset you want. 
# According to the syllabus we are looking for TCGA study on Breast Invasive Carcinoma (TCGA, Cell 2015) 
# However lets first search for all the breast cancer studies available so that we can look to see what is available. 

breast = studies[ grepl("Breast", studies$name, ignore.case = T ), ]

# ? what is the T for what if you use lower case instead of upper case

k2(breast, "list of breast related studies")

### First BONUS question! try to find if there are other cancer data you might interested in, eg Lung Cancer

# from the table we now know that the study id is:  brca_tcga_pub2015 

brca.study = "brca_tcga_pub2015"

# note: each institution has different ways of organizing their data
# ok from the study we can identify all the available data by using a case list API 


mycaselist = getCaseLists(cbiop,brca.study)


# notice that descriptions contains not only type of samples but sometimes it includes what type of data they are.  

k2 ( head ( mycaselist[ ,c("case_list_id",	 "case_list_name",	 "case_list_description",	 "cancer_study_id"	)], 50), "list of cases for the BRCA CELL 2015 study" ) 


# ? Which data do you think is most relevant for this workshop? 

# here we see that brca_tcga_pub2015_all is probably the best to use because it includes ALL Complete Tumors 
# compared to brca_tcga_pub2015_her2pos which are only samples that are Her2-positive breast tumors 
```

## For **reproducibility** we need to consider either the "freeze" case or store a local copy.  This is an important point that is often overlook!
* Also be aware that cbio is only 1 of many different portals available to the public. It is not necessarily the best.  

```{r}


case.list.id = "brca_tcga_pub2015_all"


# ok now using All Complete Tumors lets see if your sample is present. 

# first lets get all the TCGA ID associated with this case 
mysample = mycaselist[ mycaselist$case_list_id == case.list.id, ]$case_ids
mysample = unlist ( strsplit(mysample, " ") )

# ? mysample variable now contains all the TCGA samples in this study, can you check if your sample is exists? 

# here is the list of samples we stored in the auxillary. Do you see your sample in here? 

cat ( samples, sep="\n")
```

## Here I chose a random sample to test 
* IMPORTANT! here you need to use your own assigned TCGA ID so make sure that it is listed. 

```{r}

myid = "TCGA-E2-A1LE-01"
mysample[mysample== myid]

# here is another way
intersect ( mysample, myid)
# yet another 
mysample[grepl(myid, mysample)]

#? can you think of any reason to use grep over another? 

### Now lets check if it matches with what is reported: 817 
length ( mysample)

```

## With this list of TCGA ids we can also get clinical data for each of the samples. 

```{r}
# lets get clinical some data
myclinicaldata = getClinicalData(cbiop,getCaseLists(cbiop,brca.study)[1,1])
# clean up the names a bit 
myclinicaldata$pid = gsub ( "\\.", "-", row.names(myclinicaldata))
# lets get the clinical information for your TCGA id. 
k2 ( t ( myclinicaldata[myclinicaldata$pid %in% myid, ] ))

# important fields to consider. 
# AGE, AJCC_PATHOLOGIC_TUMOR_STAGE, Ethnicity
# DFS ( disease free), OS ( overall survival)
# this is quite extraordinary how much data is in here. 
# think about your experimental designs and what you can do with this! 


### Bonus Lets have some fun here, even with just a table of clinical data we can do some investigation. 
# For example how would you figure out if one of your sample ( the assigned TCGA) is a Triple-negative or HER2+. 
# hint look at the caselist and work backwards using the intersect function

k2 ( head ( mycaselist[ ,c("case_list_id",	 "case_list_name",	 "case_list_description",	 "cancer_study_id"	)], 50), "list of cases for the BRCA CELL 2015 study" ) 

t3.study = "brca_tcga_pub2015_trineg"
t3 = mycaselist[ mycaselist$case_list_id == t3.study, ]$case_ids
t3 = unlist ( strsplit(t3, " ") )


her2p.study = "brca_tcga_pub2015_her2pos"
her2p = mycaselist[ mycaselist$case_list_id == her2p.study, ]$case_ids
her2p = unlist ( strsplit(her2p, " ") )

er_neg.study = "brca_tcga_pub2015_erneg"
er_neg = mycaselist[ mycaselist$case_list_id == er_neg.study, ]$case_ids
er_neg = unlist ( strsplit(er_neg, " ") )


er_POS.study = "brca_tcga_pub2015_erpos"
er_POS = mycaselist[ mycaselist$case_list_id == er_POS.study, ]$case_ids
er_POS = unlist ( strsplit(er_POS, " ") )


# answer. 
t3[t3==myid]
her2p[her2p==myid]

# Also suppose you ask how many are HER2 positive but ER +? 

venn1 = data.frame ( her2p= length ( her2p)
                    , er_neg= length ( er_POS )
                    , int= length ( intersect(er_POS,her2p))  
                        )
names ( venn1 )[1:2] = c("Her2+","ER+") 

venn.this ( 
    venn1, 
    type = 2, cp= c("#3C8A9B" , "#9B445D"), 
    dgg=180
) 

# Now out of curiosity we want to know if there is a difference in age based on these attributes. 

age.both = intersect (er_POS,her2p )
age.both=myclinicaldata[ myclinicaldata$pid %in% age.both, ]$AGE
age.Her2only.noER = setdiff( her2p, er_POS)
age.Her2only.noER = myclinicaldata[ myclinicaldata$pid %in% age.Her2only.noER, ]$AGE

temp = rbind ( 
data.frame ( age = age.both, type= rep("both", length(age.both)) )    
,data.frame ( age = age.Her2only.noER, type= rep("Heronly", length(age.Her2only.noER)))   
)

median ( age.Her2only.noER )
median ( age.both )

ggplot( temp, aes(x=age, fill=type)) +
    geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity') +
    scale_fill_manual(values=c("#69b3a2", "#404080")) 
```

BONUS question to take home. As a pharmacologist how can you use DFS and OS? For example 
the histogram above only shows age of onset.  However, question what if you hypothesize
that having HER2 positive is much worse of survival then having both HER2 and ER+?


```{r}
 

# Now lets see what information is available for your study 
mygeneticprofile = getGeneticProfiles(cbiop,brca.study)
k2 ( mygeneticprofile, "types of data available for chosen case" )


# so looking at that we now know it contains several interesting modalities.  
# lets pick up 3 types of analysis that may be of interests 

mutation = mygeneticprofile[grepl( "Mutation", mygeneticprofile$genetic_profile_name), ]$genetic_profile_id 

# ? what other modalities do you think may be of interest
# anyone want to guess what "probe most anti-correlated with expression means"
cna = mygeneticprofile[grepl( "Putative copy", mygeneticprofile$genetic_profile_name),1 ]

exp = mygeneticprofile[grepl( "V2 RSEM", mygeneticprofile$genetic_profile_name),1 ]
exp = exp[ grepl ( "mrna$", exp)]

```



# Now that we have a basic understanding of what is available lets go and download the actual data. 
**Important** Due to the nature of the API and having so many people use it at once the connection might fail. If that is the case retry or if all else fails use the frozen version which has already been previosly downloaded. 

```{r}

# lets collect this through a loop so not to overwhelm the system
# its important to be mindful of resources when mining data.  We need to play nice and consider the server load. 
# ? what do you think we can do to miminize server load? 

# For this class we are going to just download genes that are relevant to cancer. 
# genes that had already been implicated in cancer.
# We do this by looking through different organization and cross reference a list of genes. 

k2 ( cancer.list[17:25, ], "cancer list")

length ( cancer.gene)


# lets check this list to make sure genes that are relevant to breast cancer exists 

imp = as.character ( cancer.list[grepl("^BRCA|^ATM$|^BARD1$|^CDH1$|^CHEK2$|^NBN$|^NF1$|^PALB2$|^PTEN$", cancer.list$gene), ]$gene )

# ? why did we add ^ in front and $ for only some genes and not others?

k2 ( cancer.list[cancer.list$gene %in% imp, ], "breast cancer genes")



total = ceiling ( length( cancer.gene)/100  )  *100
mutations = data.frame (   stringsAsFactors = F )
expression = data.frame (   stringsAsFactors = F )
e = 1
for(i in seq(from=150, to=total, by=150) ){
   
    if ( i > length(cancer.gene)){
        i = length(cancer.gene)
    }
    
    print ( paste ( e, i ))
    temp = getMutationData(cbiop , brca.study, mutation, cancer.gene[e:i])

    
    
    mutations = rbind ( mutations, temp)
    Sys.sleep (2) # lets give the system a break
        e2 = 1
        exp.temp2 = 1
        for(i2 in seq(from=150, to=length ( myclinicaldata$pid), by=150)){
        exp.temp = getProfileData(cbiop , cancer.gene[e:i]
                              , exp
                              , getCaseLists(cbiop,brca.study)[1,1] , myclinicaldata$pid[e2:i2])
        if ( exp.temp2 == 1){
            exp.temp2 =   t ( exp.temp)
        }else {
            exp.temp2 =  cbind ( exp.temp2,  t ( exp.temp)) 
        }
        e2 = i2+ 1 
        Sys.sleep (1) # lets give the system a break
        }
        
    e = i + 1    
    expression = rbind ( expression, exp.temp2)    
    
    
    
}
# create coordinates
mutations$igv = paste ( mutations$chr, mutations$start_position, mutations$end_position, mutations$reference_allele, mutations$variant_allele)

# there is an error in the mutation table.  The links are obsolete.  Here we fix it. 
# by replacing the older url with this: http://mutationassessor.org/r3

mutations$xvar_link = gsub('getma.org','http://mutationassessor.org/r3', mutations$xvar_link)
mutations$xvar_link_msa = gsub('getma.org','http://mutationassessor.org/r3', mutations$xvar_link_msa )
mutations$xvar_link_pdb = gsub('getma.org','http://mutationassessor.org/r3', mutations$xvar_link_pdb  )
```

# Besides Mutation data we also downloaded expression RNAseq data in the form of TPM. 
* Let take a look at some of our favorite genes. 
* Here is an example of comparing samples that test positive for HER2 vs. those that don't
* _NOTE_ ERBB2 is the HER2 gene
* EGFR is the HER1 gene 

```{r fig=TRUE,fig.width=12, fig.height=12}
expressionc= expression[complete.cases(expression), ]
samplesnice = gsub ( "-",".", samples)
expressionc[ row.names (expressionc ) %in% c(  "TP53" , "KRAS"), samplesnice ]

do.these = c( "TP53", "BRCA2","PTEN", "BRCA1", "ERBB2", "EGFR"  )
plot.out = list()

for ( gene in do.these  ){

g1 = melt (  expression[gene,  ]  )
colnames ( g1 )= c("pid","value")
g1$pid = gsub ( "\\.","-", g1$pid)

g1$HER2p = ifelse ( g1$pid %in% her2p, "HER2P", "NO.HER2")
g1$value = log2 ( g1$value + 1 )

mann_whit = wilcox.test(g1[g1$HER2p == "NO.HER2", ]$value,
                        g1[g1$HER2p != "NO.HER2", ]$value
                        ,paired=FALSE) 

p = round ( mann_whit$p.value, 2 )  
bon = p * length ( do.these )

gthis = ggplot(g1, aes(y=value, x=HER2p, fill=HER2p)) +
        geom_violin()+ 
        geom_jitter(shape=19, position=position_jitter(0.07), size=3 ) +
        theme_bw() +
        ylab("log2 (tpm + 1 ) ") +
        xlab("") +
        theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
              
              axis.text.y = element_text(size=12),
              axis.text.x = element_text(angle = 90, size=11.5),
              axis.title.x = element_text(size=22),
              
              axis.title.y     = element_text(size=22), 
              legend.text      =element_text(size=12)
        ) + stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
                         geom = "crossbar", width = .5)  + ggtitle ( gene ) +
    scale_fill_manual(values = c("#a6cee3","#b2df8a", "#fdbf6f") ) +
    ggtitle ( paste ( gene, "p.value", p, "corr", bon ))
plot.out[[gene]] = gthis
#plot ( gthis )
}

do.call("grid.arrange", c(plot.out, ncol=2))



```

# Now lets see if there are genes that correlate well with the HER2 gene 

```{r fig=TRUE,fig.width=12, fig.height=12}

exp = data.frame ( t ( log2 ( expression + 1 )))
temp = cor(exp[-1], exp$ERBB2, method="spearman") 
her2.c = as.numeric(temp )
names ( her2.c) = row.names ( temp )
her2.c = sort ( abs ( her2.c ), decreasing = T )
her2.c = round ( her2.c, 3)


do.these = head ( names ( her2.c ))
plot.out = list()

for ( gene in do.these  ){

g1 = melt (  expression[gene,  ]  )
colnames ( g1 )= c("pid","value")
g1$pid = gsub ( "\\.","-", g1$pid)

g1$HER2p = ifelse ( g1$pid %in% her2p, "HER2P", "NO.HER2")
g1$value = log2 ( g1$value + 1 )

mann_whit = wilcox.test(g1[g1$HER2p == "NO.HER2", ]$value,
                        g1[g1$HER2p != "NO.HER2", ]$value
                        ,paired=FALSE) 

p = round ( mann_whit$p.value, 2 )  
bon = p * length ( do.these )

gthis = ggplot(g1, aes(y=value, x=HER2p, fill=HER2p)) +
        geom_violin()+ 
        geom_jitter(shape=19, position=position_jitter(0.07), size=3 ) +
        theme_bw() +
        ylab("log2 (tpm + 1 ) ") +
        xlab("") +
        theme(legend.position="none", legend.title=element_blank(), legend.key = element_blank(),
              
              axis.text.y = element_text(size=12),
              axis.text.x = element_text(angle = 90, size=11.5),
              axis.title.x = element_text(size=22),
              
              axis.title.y     = element_text(size=22), 
              legend.text      =element_text(size=12)
        ) + stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
                         geom = "crossbar", width = .5)  + ggtitle ( gene ) +
    scale_fill_manual(values = c("#a6cee3","#b2df8a", "#fdbf6f") ) +
    ggtitle ( paste ( gene, "p.value", p, "corr", bon, "spearm", her2.c[gene] ))
plot.out[[gene]] = gthis
#plot ( gthis )
}

do.call("grid.arrange", c(plot.out, ncol=2))

```


# study sample id and see if there are any alteration in expression

```{r}

scaled.dat <- data.frame ( t ( scale(exp) )) # note that I had to use the log2 version
myid2 = gsub ( "-", ".", myid )
my.dat = scaled.dat[, myid2, drop=T] 
names ( my.dat)= row.names (scaled.dat )
my.dat = sort ( abs ( my.dat ), decreasing = T )
my.dat = round ( my.dat, 3)

gene = names (  my.dat[1] )
g1 = melt (  log2 ( expression [gene,   ] + 1 )   )

colnames ( g1 )= c("pid","value")
g1 = g1[ order ( g1$value), ]
g1$pid = factor ( g1$pid, levels=g1$pid)

ggplot(g1, aes(x=pid, y=value, color=  value  )) +
  geom_bar(stat="identity") +
 scale_fill_brewer(palette = "Set1")    +
  xlab("") + ylab("") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()

  ) + theme_void()  + 
  geom_text_repel(
        data = g1[g1$pid == myid2,  ], 
        aes(
          label = pid
        )
        , 
        fontface="bold", 
        color="black",
        size = 5, 
        nudge_x = 20.15,
    box.padding = 1.5,
    nudge_y = -.5,
  
      ) + ggtitle ( paste ( gene )  )




```

# lets see if there are any druggable gene expression 

```{r}
# list of "overexpressed" genes for id 
my.dat[my.dat> 1.96 ]

drug.up[ drug.up$gene %in% names ( my.dat[my.dat> 1.96 ] ) , ]

```


# lets study the mutation table

```{r}

dim ( mutations )

# lets take a few minutes here to go over the different fields. 
names ( mutations )


unique( mutations$mutation_type)

# ? based on the lecture previously can identify the different mutation_type
# ? based on the lecture previously can you figure out what each of these mean?  
# for example which of these are nonsynonymous substitution?
# ? for druggability which one of these do you think its more likely to be useful

# ? can you guess what is functional_impact_score
k2 ( data.frame ( table ( mutations$functional_impact_score)  ), "impact scores " )

# make sure that your sample has mutations. 
#? if you don't see any or obvious mutation where else can look at for the driver signal? 

nrow ( mutations [ mutations$case_id ==  myid , ] )

# ? if you have more than one sample, say you want to check all 10 samples in the class, how would you do this? 
# ? this is important, for example previously we were able to subset for Triple-negative breast tumors

samples

final =  mutations [ mutations$case_id %in% samples, ]
unique ( final$case_id)

# so as you can see above there are a total of 10 unique samples. 

# lets clear up junk 
gc()
                          

```


# Now lets talk a bit about how to identify pathogenic mutations. 
* recall some of the things that were mentioned previously? 

```{r}


# For example we can detect important mutations is to see if other resources are available. 
# Here we use cosmic a manaully curated list of recurrent mutations in cancer and look for keyword breast 

k2 ( head ( cosmic.70[ grepl("breast",  cosmic.70$description ), ] ), "breast cancer in cosmic")

# lets integrate cosmic into our mutation table

mutations = merge ( mutations, cosmic.70, by="igv", all.x=T, all.y=F )

# ? note how we are merging by coordinates.  This is the most accurate way to do this. 
# ? why do you think all.x = T and all.y=F 

# let us find out how many mutations we have that is also found in cosmic 
# the description field should not be empty if there was a cosmic entry

nrow ( mutations[ ! is.na ( mutations$description ),  ] ) / nrow (mutations)

# ? wow thats a huge ratio , why do think the ratio is so high? 

# BONUS check if cosmic contain BRCA mutations.  This is a good sanity check. 

k2 ( head ( mutations[ grepl("^BRCA", mutations$gene_symbol), c("amino_acid_change","chr.x","start_position","end_position","reference_allele", "variant_allele",
"gene_symbol","case_id","mutation_status","mutation_type","description")] ) )
# BONUS go to https://cancer.sanger.ac.uk/cosmic and compare this. 
# ? does everything match why not? 



# BONUS can you find all BRCA mutation that is Missense_Mutation AND found is cosmic in the mutation db?
# hint, mutations$mutation_type == "Missense_Mutation"
```

# Now that the data has been succesfully donwloaded lets study your breast cancer set. 

```{r}



# lets tabulate the types of mutations  

mutation.type = data.frame ( table ( mutations$mutation_type))

mutation.type = mutation.type[ order ( mutation.type$Freq), ]

mutation.type$Var1 = factor ( mutation.type$Var1, levels = mutation.type$Var1)

ggplot(mutation.type, aes(Var1,Freq), label=Freq ) +
    geom_bar(aes(fill = Freq), stat="identity", position = "dodge") +
    coord_flip() +
    scale_fill_distiller(palette = "RdBu") + xlab("") + ylab("") +
    theme(strip.text.y = element_text(angle = 0), legend.position="none") +
    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=.4, hjust = .5, size=5) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"), 
          text = element_text(size=16),  # size of label
          axis.text.x = element_text(angle=0, hjust=1) ) + ggtitle ( "All mutation types")


# As mentioned previously one reliable way to look for relevant mutations is to match it against what has been previously observed and/or study. Here we can try to use Cosmic however other organization may be relevant as well including clinvar

cosmic.mutation = mutations [ !is.na(mutations$description), ]

mutation.type = data.frame ( table ( cosmic.mutation$mutation_type))

mutation.type = mutation.type[ order ( mutation.type$Freq), ]
mutation.type$Var1 = factor ( mutation.type$Var1, levels = mutation.type$Var1)

ggplot(mutation.type, aes(Var1,Freq), label=Freq ) +
    geom_bar(aes(fill = Freq), stat="identity", position = "dodge") +
    coord_flip() +
    scale_fill_distiller(palette = "RdBu") + xlab("") + ylab("") +
    theme(strip.text.y = element_text(angle = 0), legend.position="none") +
    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=.4, hjust = .5, size=5) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"), 
          text = element_text(size=16),  # size of label
          axis.text.x = element_text(angle=0, hjust=1) ) + ggtitle ( "Mutation types found in cosmic")


# ? seems like missense mutation is the most recurrent.  Can you think of any way to filter the data further 

# ? in the lecture we talked about depth.  So lets calculate this. 

mutations$dp = mutations$variant_read_count_tumor + mutations$reference_read_count_tumor 
quantile ( mutations$dp)

 ggplot(mutations, aes(x=dp)) + 
  geom_density( color="darkblue", fill="steelblue" ) + geom_vline(xintercept = 21, linetype="dotted", 
                color = "grey", size=1.5) + ggtitle ( " density plot, total depth")

# ? by looking at this plot and quantile we can see that there is a good amount of mutations that are under 21 
# ? how would you remove this?

# before  
dim ( mutations )
# after 
mutations = mutations[ mutations$dp > 21, ]
dim (mutations )


# another thing we can look at is Allele frequency
# lets calculate allele freqeuncy here as well 

mutations$af = mutations$variant_read_count_tumor / mutations$dp

 ggplot(mutations, aes(x=af)) + 
  geom_density( color="darkblue", fill="#e8975a" ) + geom_vline(xintercept = .1, linetype="dotted", 
                color = "grey", size=1.5) + ggtitle ( " density plot, allele freqeuncy")

# ? what do you think it means when af is higher for a particular mutation 
# Bonus pick the top 10 mutations ranked by af 

mutations = mutations[ order ( - mutations$af ) , ]

# clean mut head
cleanmut = c ("igv","dp","af","gene_symbol","case_id",	"mutation_type", "amino_acid_change" , "functional_impact_score"
              , "variant_read_count_tumor"	,"reference_read_count_tumor" , "description" ) 

k2 ( head ( mutations[ , cleanmut], 10) , "top 10 variant ranked by AF")

# lets further filter the mutations with any af > .1
mutations = mutations[ mutations$af > .1, ]
dim (mutations)


# plot 20 highest recurrent gene mutations

mutation.gene = data.frame ( table ( mutations$gene_symbol))

mutation.gene = mutation.gene[ order ( - mutation.gene$Freq), ]
mutation.gene$Var1 = factor ( mutation.gene$Var1, levels = rev ( mutation.gene$Var1) )

ggplot(mutation.gene [ 1:20, ] , aes(Var1,Freq), label=Freq ) +
    geom_bar(aes(fill = Freq), stat="identity", position = "dodge") +
    coord_flip() +
    scale_fill_distiller(palette = "RdBu") + xlab("") + ylab("") +
    theme(strip.text.y = element_text(angle = 0), legend.position="none") +
    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=.4, hjust = .5, size=5) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"), 
          text = element_text(size=16),  # size of label
          axis.text.x = element_text(angle=0, hjust=1) ) + ggtitle ( "Most recurrent genes")







# based on literature these are some of the genes involved in breast cancer. 
imp 

# lets study them and see what types of mutations corresponds most with these genes. 

mutation.brca = data.frame ( table ( mutations[ mutations$gene_symbol %in% imp, ] $mutation_type))

mutation.brca = mutation.brca[ order ( mutation.brca$Freq), ]
mutation.brca$Var1 = factor ( mutation.brca$Var1, levels = mutation.brca$Var1)

ggplot(mutation.brca, aes(Var1,Freq), label=Freq ) +
    geom_bar(aes(fill = Freq), stat="identity", position = "dodge") +
    coord_flip() +
    scale_fill_distiller(palette = "RdBu") + xlab("") + ylab("") +
    theme(strip.text.y = element_text(angle = 0), legend.position="none") +
    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=.4, hjust = .5, size=5) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"), 
          text = element_text(size=16),  # size of label
          axis.text.x = element_text(angle=0, hjust=1) ) + ggtitle ( "Mutation-types in genes associated with breast cancer")

# lets assume that we don't know these genes 
# BONUS can you think of another way to rediscover these genes and/or find novel ones?

# HERE is where the fun is! The discovery part.  And there are many many ways to do this.  
# so be creative! 

# here is a 2 sec way to do this. 
# simply look in cosmic to see how often these genes appears in breast cancer. 
imp.brac = data.frame ( table ( mutations[ grepl( "breast", mutations$description), ]$gene_symbol ) )
imp.brac$fraction = imp.brac$Freq / nrow ( mutations)
imp.brac = imp.brac [ order ( -imp.brac$fraction) , ]

k2 ( head ( imp.brac, 10), "top cosmic genes found in breast cancer")

# ? how does this compare to our plot above 
# both analysis seem to show a lot of overlap eg that TP53 and PIK3CA are important.  Lets breakdown the mutation type a bit more for the top 5

top5 = as.character ( imp.brac$Var1[1:5] )
top5 = mutations[ mutations$gene_symbol %in% top5, ]$amino_acid_change
top5 = data.frame ( table ( top5 ) )
top5 = top5 [ order ( - top5$Freq), ]

# quick hack to add url, will not work with *

top5$url = paste0(url,top5$top5)
k2 ( head ( top5, 20), "top 20 recurrent mutations")
# take the first mutation and google it what does it say? 
# take the second, what does google say? 



# part of studying mutations is look for clinical significance.  One way to do this is to match your mutations to actionable 
# targets


mutations$drug = paste ( mutations$gene_symbol, mutations$amino_acid_change)

action = drug[ drug$match %in% unique ( mutations$drug  ), ]
action.f = c("gene" ,"variant" , "disease", "drugs", "evidence_type", 
             "clinical_significance", "evidence_statement", "evidence_civic_url", "gene_civic_url" , "match"
)
k2 ( head ( action[ ,action.f], 20) ,"example of actionable mutations in the BRCA set" )

# Bonus can you figure out which samples have druggable targets?  
#? look at evidence type. Can yo figure out which of these are diagnostic vs targetable? 

k2 ( head ( action[action$clinical_significance == "Sensitivity/Response" ,action.f], 20) )



```




# HW at this point please use your own TCGA ID, study it using what you learn above.  Assess QC on your sample, study what genes are most common and which if any are targetable. 

```{r}
# here is example where I chose, myid = "TCGA-C8-A131-01"

 
actionid = drug[ drug$match %in% unique ( mutations[mutations$case_id == myid, ]  $drug  ), ]

actionid = actionid %>% dplyr::group_by(variant) %>%
        dplyr::summarise(
            gene = paste(unique ( gene ), collapse = "," ) ,
            disease = paste(unique ( disease ), collapse = "," ),
            clinical_significance =  paste(unique ( clinical_significance ), collapse = "," ),
            drugs = paste(unique ( drugs ), collapse = "," ) 
            
        ) %>%
        data.frame()

k2( actionid, myid)
 
```




